zookeeper/zookeeper.go

```diff
@@ -0,0 +1,204 @@
+		opts.ServerPathFormat = ZookeeperDefaultServerPath
+	}
+	if opts.ClientPathFormat == "" {
+		opts.ClientPathFormat = ZookeeperDefaultClientPath
+	}
+
+	conn, _, err := zk.Connect(opts.Server, time.Second*5, zk.WithLogger(opts.CustomLogger))
+	if err != nil {
+		return nil, err
+	}
+	prefixTemplate, err := template.New("prefix").Parse(opts.Prefix)
+	if err != nil {
+		return nil, err
+	}
+	serverPathTemplate, err := template.New("serverPathID").Parse(opts.ServerPathFormat)
+	if err != nil {
+		return nil, err
+	}
+	clientPathTemplate, err := template.New("clientPathID").Parse(opts.ClientPathFormat)
+	if err != nil {
+		return nil, err
+	}
+	c := &client{
+		zConn:              conn,
+		parser:             opts.ConfigParser,
+		serverPathTemplate: serverPathTemplate,
+		clientPathTemplate: clientPathTemplate,
+		prefixTemplate:     prefixTemplate,
+	}
+	return c, nil
+}
+
+// SetParser support customise parser
+func (c *client) SetParser(parser ConfigParser) {
+	c.parser = parser
+}
+
+func (c *client) render(cpc *ConfigParamConfig, t *template.Template) (string, error) {
+	var tpl bytes.Buffer
+	err := t.Execute(&tpl, cpc)
+	if err != nil {
+		return "", err
+	}
+	return tpl.String(), nil
+}
+
+// ServerConfigParam render server config parameters
+func (c *client) ServerConfigParam(cpc *ConfigParamConfig) (ConfigParam, error) {
+	return c.configParam(cpc, c.serverPathTemplate)
+}
+
+// ClientConfigParam render client config parameters
+func (c *client) ClientConfigParam(cpc *ConfigParamConfig) (ConfigParam, error) {
+	return c.configParam(cpc, c.clientPathTemplate)
+}
+
+// configParam render config parameters. All the parameters can be customized with CustomFunction.
+// ConfigParam explain:
+//  1. Prefix: /KitexConfig by default.
+//  2. ServerPath: {{.ServerServiceName}}.{{.Category}} by default.
+//     ClientPath: {{.ClientServiceName}}.{{.ServerServiceName}}.{{.Category}} by default.
+func (c *client) configParam(cpc *ConfigParamConfig, t *template.Template) (ConfigParam, error) {
+	param := ConfigParam{}
+	var err error
+	param.Path, err = c.render(cpc, t)
+	if err != nil {
+		return param, err
+	}
+	param.Prefix, err = c.render(cpc, c.prefixTemplate)
+	if err != nil {
+		return param, err
+	}
+	return param, nil
+}
+
+// DeregisterConfig deregister the config.
+func (c *client) DeregisterConfig(path string, uniqueID int64) {
+	mu.Lock()
+	clientKey := path + "/" + strconv.FormatInt(uniqueID, 10)
+	cancel := ctxMap[clientKey]
+	cancel()
+	mu.Unlock()
+}
+
+// RegisterConfigCallback register the callback function to zookeeper client.
+func (c *client) RegisterConfigCallback(ctx context.Context, path string, uniqueID int64, callback func(string, ConfigParser)) {
+	clientCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		mu.Lock()
+		clientKey := path + "/" + strconv.FormatInt(uniqueID, 10)
+		ctxMap[clientKey] = cancel
+		mu.Unlock()
+		for {
+			_, _, watchChan, err := c.zConn.ExistsW(path)
+			if err != nil {
```

Comment:
> In `config-etcd`, the watchChan is obtained before for loop. Is it different for zk here?

Once watchChan is triggered, it will be deleted, so the Watch needs to be reset every time a Watch notification is received.

Reply:
> What happens it the zk server is temporarily unavailable?

The server keeps trying to reconnect to zookeeper

Reply:
You are using the declaration of a temporary variable:=

Reply:
New a goroutine func maybe better for read

Reply:
> You are using the declaration of a temporary variable:=

I see. I put the variable outside the loop for reuse